protocol/lavaprotocol/finalization_consensus.go

```diff
@@ -100,83 +99,113 @@ func (fc *FinalizationConsensus) insertProviderToConsensus(blockDistanceForFinal
 // create new consensus group if no consensus matched
 // check for discrepancy with old epoch
 // checks if there is a consensus mismatch between hashes provided by different providers
-func (fc *FinalizationConsensus) UpdateFinalizedHashes(blockDistanceForFinalizedData int64, providerAddress string, finalizedBlocks map[int64]string, req *pairingtypes.RelaySession, reply *pairingtypes.RelayReply) (finalizationConflict *conflicttypes.FinalizationConflict, err error) {
+func (fc *FinalizationConsensus) UpdateFinalizedHashes(blockDistanceForFinalizedData int64, consumerAddress sdk.AccAddress, providerAddress string, finalizedBlocks map[int64]string, relaySession *pairingtypes.RelaySession, reply *pairingtypes.RelayReply) (finalizationConflict *conflicttypes.FinalizationConflict, err error) {
+	logSuccessUpdate := func() {
+		if debug {
+			utils.LavaFormatDebug("finalization information update successfully",
+				utils.LogAttr("specId", fc.specId),
+				utils.LogAttr("finalizationData", finalizedBlocks),
+				utils.LogAttr("currentBlockToHashesToAgreeingProviders", fc.currentBlockToHashesToAgreeingProviders),
+			)
+		}
+	}
+
 	latestBlock := reply.LatestBlock
+
 	fc.providerDataContainersMu.Lock()
-	defer fc.providerDataContainersMu.Unlock()
+	defer func() {
+		fc.insertProviderToConsensus(latestBlock, blockDistanceForFinalizedData, finalizedBlocks, reply, relaySession, providerAddress)
+		fc.providerDataContainersMu.Unlock()
+	}()
 
-	if len(fc.currentProviderHashesConsensus) == 0 && len(fc.prevEpochProviderHashesConsensus) == 0 {
-		newHashConsensus := fc.newProviderHashesConsensus(blockDistanceForFinalizedData, providerAddress, latestBlock, finalizedBlocks, reply, req)
-		fc.currentProviderHashesConsensus = append(make([]ProviderHashesConsensus, 0), newHashConsensus)
+	var blockToHashesToAgreeingProviders BlockToHashesToAgreeingProviders
```

Comment:
this copies the entire map and duplicates it. use a pointer instead.

Reply:
Not relavent anymore